### PR TITLE
Adds TypeScript

### PR DESCRIPTION
**Pull request by artem-zakharchenko**
#### :rocket: Why this change?

See original motivation in #1497 

#### :memo: Related issues and Pull Requests

- More granular introduction of #1497 
- Implements #1524 

#### :white_check_mark: What didn't I forget?

- [x] To write docs
- [x] To write tests
- [x] To put [Conventional Changelog](https://dredd.org/en/latest/internals.html#sem-rel) prefixes in front of all my commits and run `npm run lint`

### CODE REVIEW

**Comment by artem-zakharchenko** [comment]
As the first step I move `lib/options.json` from the `lib` directory to the root directory.

This is necessary due to how TypeScript compiles the source code directory. Basically, it will output the built code in to the dist folder, but it won't process JSON files by default. There are two options how to approach JSON processing:

## Set `resolveJsonModule: true`

With this option set in `tsconfig.json`, TypeScript will treat JSON files imports as a regular dependency and will compile JSON files into dist directory alongside the other built modules.

**The issue:** I haven't found a way to exclude certain JSON files (tried `files`, `exclude`, globs), which results into TypeScript also compiling `package.json`, as Dredd imports in source code. Because of the root-level JSON dependency TypeScript recreates the entire directory structure starting from the most distant dependency, which results into the following folder structure upon build:

```
- build
  - package.json // gets copied
  - lib/ // here's the source directory, copied
    - Dredd.js
- lib/
  - Dredd.js
- package.json
```

## Do not process JSON

Move JSON files into the directory that would have the same import paths for both source and dist files (which means their parent directory). This way the build directory structure represents the source code directory structure 1-1:

```
- build/
  - Dredd.js
- lib/
  - Dredd.js
- options.json
- package.json
```

I find the second option to be the lesser evil and, perhaps, even making `options.json` more evident to be a global source of truth for source code and tests.

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  "compileOnSave": true,
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es2017",
+    "lib": ["es2017"],
+    "typeRoots": ["node_modules/@types"],
+    "rootDir": "./lib",
+    "outDir": "./build",
+    "moduleResolution": "node",
+    "esModuleInterop": true,
```
Comment (artem-zakharchenko) [inline tsconfig.json]: Fixes the imports of ES modules that have the "default" export. Doesn't require to write:

```js
const foo = require('module').default
```

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "allowJs": true,
```
Comment (artem-zakharchenko) [inline tsconfig.json]: Configuring JS compatibility mode for now, as there is no intention to rewrite the entire Dredd to TypeScript in a single pull request.

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    "allowSyntheticDefaultImports": true
```
Comment (artem-zakharchenko) [inline tsconfig.json]: Preserves default module resolution, so you don't have to write:

```js
import * as path from 'path'
```

> As the `path` module has no default export, TypeScript would require you to import its exports using wildcard. While that's a valid code, I'd lean toward preserving as much backward compatibility as possible at this early stage of adoption.

**Comment by artem-zakharchenko** [comment]
The next step is to add a basic TypeScript configuration and introduce a build step.

First step does not contain any actual modules written in TypeScript, so it acts as a proxy of the existing source code. It does, however, allow to write the source code in ES6, which I'm about to add (mainly `require` -> `import`). **The source code compilation is at place**.

File: appveyor.yml [hunk 4]
```diff
@@ -1,15 +1,16 @@
 environment:
-  nodejs_version: "10"
```
Comment (honzajavorek) [inline appveyor.yml]: This is 12 now after merging https://github.com/apiaryio/dredd/pull/1495

Comment (artem-zakharchenko) [reply appveyor.yml]: Will be updated after this branch is rebased against the latest `master`.

**Comment by artem-zakharchenko** [comment]
Preparing the next chunk of changes together with mocha+ts setup, as rewritten to ES6 source code won't be digested by the test suites, resulting into failed CI. To prevent this, I will push both source code rewrite and mocha+ts setup altogether (in separate commits).

File: lib/index.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import Dredd from './Dredd';
+
+module.exports = Dredd;
```
Comment (artem-zakharchenko) [inline lib/index.js]: I'm adding explicit entry module so it can be exported properly.

By default, TypeScript will compile the source code into CJS, but preserve the default/named exports. This means that if you wish to import a distributed Dredd you cannot do:

```js
const Dredd = require('dredd')
```

You would have to do:

```js
const Dredd = require('dredd').default
```

To solve this, an explicit entry helps. It exports the default export as `module.exports`, allowing to preserve the same require/import as before.

**Comment by artem-zakharchenko** [comment]
Tests adjusted to run with source code being in ES6/TypeScript.

That's achieved using `ts-node` module that's required by `mocha` before test run. Tests use their own `tsconfig.json` extending the default one used for the library's build. There are a few quirks about imports and proxyquring, refer to the changed files for more details.

File: lib/index.js (hunk 5 above)
Comment (honzajavorek) [reply lib/index.js]: This should probably be a comment in code.

File: test/.eslintrc.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
 module.exports = {
   env: {
-    mocha: true,
-    node: true,
```
Comment (honzajavorek) [inline test/.eslintrc.js]: Why are these changes needed?

Comment (artem-zakharchenko) [reply test/.eslintrc.js]: Left-overs from copying the test directory. Fixed pushed in e895117

File: lib/index.js (hunk 5 above)
Comment (artem-zakharchenko) [reply lib/index.js]: Added in 2f1a89d.
